File: archive/versions/pre_fin_0_1_3/unpromoted_active_tree/apps/workbench/frontend/vite/src/api/cases.ts
export const CASES_PATH = "/api/v1/cases";

export type CreateCaseDraftCommand = {
  query: string;
  as_of: string;
  language: string;
  source_policy_ref: string;
  idempotency_key: string;
};

export type TaskCenterRow = {
  case_id: string;
  case_version: number;
  query: string;
  status: string;
  updated_at: string;
};

export type TaskCenterProjection = {
  items: TaskCenterRow[];
  next_cursor: string | null;
};

export type CaseWorkspaceProjection = {
  case_id: string;
  case_version: number;
  summary_version: number;
  query: string;
  as_of: string;
  language: string;
  planning_checkpoint_state: string;
};

export type FixtureCaseContext = {
  tenantId: string;
  projectId: string;
  actorId: string;
  permissions: readonly string[];
};

type ErrorEnvelope = {
  error?: {
    error_code?: string;
    message?: string;
    status_code?: number;
    trace_id?: string;
  };
  detail?: unknown;
};

export class CaseApiError extends Error {
  readonly code: string;
  readonly statusCode: number;
  readonly traceId: string | undefined;
  readonly detail: unknown;

  constructor(input: { code: string; message: string; statusCode: number; traceId?: string; detail?: unknown }) {
    super(input.message);
    this.name = "CaseApiError";
    this.code = input.code;
    this.statusCode = input.statusCode;
    this.traceId = input.traceId;
    this.detail = input.detail;
  }
}

export const DEFAULT_FIXTURE_CASE_CONTEXT: FixtureCaseContext = {
  tenantId: "fixture_internal",
  projectId: "workbench_internal",
  actorId: "analyst_internal",
  permissions: ["case:read", "case:create"],
};

export class CaseApiClient {
  constructor(private readonly context: FixtureCaseContext = DEFAULT_FIXTURE_CASE_CONTEXT) {}

  async createCase(command: CreateCaseDraftCommand): Promise<CaseWorkspaceProjection> {
    return this.request<CaseWorkspaceProjection>(CASES_PATH, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(command),
    });
  }

  async listCases(): Promise<TaskCenterProjection> {
    return this.request<TaskCenterProjection>(CASES_PATH);
  }

  async getCase(caseId: string, expectedCaseVersion?: number): Promise<CaseWorkspaceProjection> {
    const headers = new Headers();
    if (expectedCaseVersion !== undefined) {
      headers.set("X-Fin-Case-Expected-Version", String(expectedCaseVersion));
    }
    return this.request<CaseWorkspaceProjection>(`${CASES_PATH}/${encodeURIComponent(caseId)}`, { headers });
  }

  private async request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const headers = new Headers(init.headers);
    headers.set("Accept", "application/json");
    headers.set("X-Fin-Case-Tenant", this.context.tenantId);
    headers.set("X-Fin-Case-Project", this.context.projectId);
    headers.set("X-Fin-Case-Actor", this.context.actorId);
    headers.set("X-Fin-Case-Permissions", this.context.permissions.join(","));

    const response = await fetch(path, { ...init, headers });
    const payload = (await response.json()) as unknown;
    if (!response.ok) {
      throw toCaseApiError(response.status, payload);
    }
    return payload as T;
  }
}

function toCaseApiError(statusCode: number, payload: unknown): CaseApiError {
  const envelope = isErrorEnvelope(payload) ? payload : {};
  const error = envelope.error;
  return new CaseApiError({
    code: error?.error_code ?? `http_${statusCode}`,
    message: error?.message ?? "Case request failed.",
    statusCode,
    traceId: error?.trace_id,
    detail: envelope.detail,
  });
}

function isErrorEnvelope(value: unknown): value is ErrorEnvelope {
  return typeof value === "object" && value !== null;
}

File: archive/versions/pre_fin_0_1_3/unpromoted_active_tree/apps/workbench/backend/api/v1/cases.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any

from fastapi import APIRouter, Header, HTTPException, Request, Response, status
from pydantic import BaseModel, ConfigDict, Field

from ...application.case_service import CasePrincipal, CaseService, CaseServiceError, CreateCaseDraft
from sec_agent.workbench.api_contracts import request_trace_id


class CreateCaseDraftCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    query: str = Field(min_length=1)
    as_of: datetime
    language: str = Field(min_length=1)
    source_policy_ref: str = Field(min_length=1)
    idempotency_key: str = Field(min_length=1)


class TaskCenterRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    case_id: str
    case_version: int = Field(ge=0)
    query: str
    status: str
    updated_at: datetime


class TaskCenterProjection(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[TaskCenterRow]
    next_cursor: str | None = None


class CaseWorkspaceProjection(BaseModel):
    model_config = ConfigDict(extra="forbid")

    case_id: str
    case_version: int = Field(ge=0)
    summary_version: int = Field(ge=0)
    query: str
    as_of: datetime
    language: str
    planning_checkpoint_state: str


def build_cases_router(service: CaseService) -> APIRouter:
    router = APIRouter(tags=["point02-cases"])

    @router.post("/cases", status_code=status.HTTP_202_ACCEPTED, response_model=CaseWorkspaceProjection)
    def create_case(
        command: CreateCaseDraftCommand,
        request: Request,
        response: Response,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.create_case(
                CreateCaseDraft(**command.model_dump()),
                _principal(tenant_id, project_id, actor_id, permissions),
                trace_id=request_trace_id(request),
            )
        except CaseServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _case_etag(projection["case_version"])
        return projection

    @router.get("/cases", response_model=TaskCenterProjection)
    def list_cases(
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            return service.list_cases(_principal(tenant_id, project_id, actor_id, permissions))
        except CaseServiceError as exc:
            _raise_service_error(exc)

    @router.get("/cases/{case_id}", response_model=CaseWorkspaceProjection)
    def get_case(
        case_id: str,
        response: Response,
        expected_case_version: Annotated[int | None, Header(alias="X-Fin-Case-Expected-Version")] = None,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.get_case(
                case_id,
                _principal(tenant_id, project_id, actor_id, permissions),
                expected_case_version=expected_case_version,
            )
        except CaseServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _case_etag(projection["case_version"])
        return projection

    return router


def _principal(tenant_id: str | None, project_id: str | None, actor_id: str | None, permissions: str | None) -> CasePrincipal:
    return CasePrincipal(
        tenant_id=(tenant_id or "").strip(),
        project_id=(project_id or "").strip(),
        actor_id=(actor_id or "").strip(),
        permissions=frozenset(item.strip() for item in (permissions or "").split(",") if item.strip()),
    )


def _raise_service_error(error: CaseServiceError) -> None:
    raise HTTPException(status_code=error.status_code, detail=error.detail) from error


def _case_etag(version: int) -> str:
    return f'"case-version={version}"'
